src/app/3d/qgs3dmaptoolmeasureline.cpp

```diff
@@ -0,0 +1,304 @@
+};
+
+void Qgs3DMapToolMeasureLinePickHandler::handlePickOnVectorLayer( QgsVectorLayer *vlayer, QgsFeatureId id, const QVector3D &worldIntersection, Qt3DRender::QPickEvent *event )
+{
+  if ( event->button() == Qt3DRender::QPickEvent::LeftButton )
+  {
+    if ( mMeasureLineTool->done() )
+    {
+      mMeasureLineTool->mDialog->restart();
+    }
+
+    // Left button, keep addinng point
+    qInfo() << "handlePickOnVectorLayer" ;
+    QgsVector3D mapCoords = Qgs3DUtils::worldToMapCoordinates(
+                              QgsVector3D( worldIntersection.x(),
+                                           worldIntersection.y(),
+                                           worldIntersection.z() ), mMeasureLineTool->mCanvas->map()->origin() );
+    QgsPoint pt( mapCoords.x(), mapCoords.y(), mapCoords.z() );
+    qInfo() << "Coord (handlePickOnVectorLayer): " << pt.x() << " " << pt.y() << " " << pt.z();
+
+    mMeasureLineTool->mDone = false;
+    mMeasureLineTool->addPoint( pt );
+    mMeasureLineTool->mDialog->show();
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::RightButton )
+  {
+    // Finish measurement
+    qInfo() << "Finish measurement";
+    mMeasureLineTool->mDone = true;
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::MiddleButton )
+  {
+    qInfo() << "Undo - middle button";
+    mMeasureLineTool->undo();
+
+  }
+}
+
+Qgs3DMapToolMeasureLine::Qgs3DMapToolMeasureLine( Qgs3DMapCanvas *canvas )
+  : Qgs3DMapTool( canvas )
+{
+  qInfo() << "Constructed";
+  connect( mCanvas->scene(), &Qgs3DMapScene::terrainEntityChanged, this, &Qgs3DMapToolMeasureLine::onTerrainEntityChanged );
+  mPickHandler.reset( new Qgs3DMapToolMeasureLinePickHandler( this ) );
+
+  // Line style
+  mLineSymbol = new QgsLine3DSymbol;
+  mLineSymbol->setRenderAsSimpleLines( true );
+  mLineSymbol->setWidth( 4 );
+  QgsPhongMaterialSettings phongMaterial;
+  phongMaterial.setAmbient( Qt::yellow );
+  phongMaterial.setDiffuse( Qt::green );
+  mLineSymbol->setMaterial( phongMaterial );
+
+  mLineSymbolRenderer = new QgsVectorLayer3DRenderer( mLineSymbol );
+
+  // Dialog
+  mDialog = new Qgs3DMeasureDialog( this );
+  mDialog->setWindowFlags( mDialog->windowFlags() | Qt::Tool );
+  mDialog->restorePosition();
+}
+
+Qgs3DMapToolMeasureLine::~Qgs3DMapToolMeasureLine() = default;
+
+void Qgs3DMapToolMeasureLine::activate()
+{
+  qInfo() << "Measure line activated";
+  if ( QgsTerrainEntity *terrainEntity = mCanvas->scene()->terrainEntity() )
+  {
+    connect( terrainEntity->terrainPicker(), &Qt3DRender::QObjectPicker::clicked, this, &Qgs3DMapToolMeasureLine::onTerrainPicked );
+  }
+
+  mCanvas->scene()->registerPickHandler( mPickHandler.get() );
+
+  // Initialize the measurement line
+  mMeasurementLine = new QgsLineString();
+  mMeasurementLine->addZValue();
+
+  // Initialize measurement feature
+  mMeasurementFeature = new QgsFeature( QgsFeatureId( 1 ) );
+  mMeasurementFeature->setGeometry( QgsGeometry( mMeasurementLine ) );
```

Comment:
I think we do not need mMeasurementLine and mMeasurementFeature as member variables. They could be just created as local variables.

We also need to take care of object ownership:
- mMeasurementLine gets owned by QgsGeometry, so it is fragile to actually keep long-term pointer to it (may cause crash if the geometry is replaced which may cause premature destruction of the linestring and then crashes
- mMeasurementFeature leaks - it is just copied in addFeature()

Reply:
Done